WooCommerce/Classes/Authentication/AuthenticationManager.swift

```diff
@@ -449,15 +449,42 @@ private extension AuthenticationManager {
                              with credentials: AuthenticatorCredentials,
                              in navigationController: UINavigationController,
                              onDismiss: @escaping () -> Void) {
-        let viewModel = JetpackErrorViewModel(siteURL: siteURL, onJetpackSetupCompletion: { [weak self] in
-            self?.currentSelfHostedSite = nil
-            self?.presentLoginEpilogue(in: navigationController, for: credentials, onDismiss: onDismiss)
+        let viewModel = JetpackErrorViewModel(siteURL: siteURL, onJetpackSetupCompletion: { [weak self] authorizedEmailAddress in
+            if let site = self?.currentSelfHostedSite {
+                self?.currentSelfHostedSite = WordPressComSiteInfo.updateJetpackInfo(for: site)
+            }
+            guard credentials.wpcom != nil else {
+                return WordPressAuthenticator.showLoginForJustWPCom(from: navigationController, jetpackLogin: true, connectedEmail: authorizedEmailAddress)
+            }
+            // tries re-syncing to get an updated store list
+            // then attempts to present epilogue again
+            ServiceLocator.stores.synchronizeEntities { [weak self] in
+                self?.presentLoginEpilogue(in: navigationController, for: credentials, onDismiss: onDismiss)
+            }
         })
         let installJetpackUI = ULErrorViewController(viewModel: viewModel)
         navigationController.show(installJetpackUI, sender: nil)
     }
 }
 
+private extension WordPressComSiteInfo {
+    /// Copies the information from the specified site and update all Jetpack-related properties for a valid Jetpack connection.
+    ///
+    static func updateJetpackInfo(for site: WordPressComSiteInfo) -> WordPressComSiteInfo {
+        .init(remote: ["name": site.name,
+                       "description": site.tagline,
+                       "urlAfterRedirects": site.url,
+                       "hasJetpack": true,
+                       "isJetpackActive": true,
+                       "isJetpackConnected": true,
+                       "icon.img": site.icon,
+                       "isWordPressDotCom": site.isWPCom,
+                       "isWordPress": site.isWP,
+                       "exists": site.exists]
+        )
```

Comment:
nit: do you think we can add an initializer for `WordPressComSiteInfo` with typed parameters? 😅 or a unit test on this conversion?

Reply:
I removed this part following 22081054be0f1c8bb20752a7dd0e90ef0671fb50.